### PR TITLE
mining stats

### PR DESCRIPTION
**Pull request by jamesvan2019**
```bash
./cli.sh -p 18131 miningstats
```

### CODE REVIEW

**Review by lochjin — APPROVED** [review]

File: services/miner/api.go [hunk 1]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+		LastSubmit:     time.Now(),
+		LastGBTTime:    time.Now(),
+		Last100Gbts:    make([]int64, 0),
+		Last100Submits: make([]int64, 0),
+	}}
 	return pmAPI
 }
 
-//func (api *PublicMinerAPI) GetBlockTemplate(request *mining.TemplateRequest) (interface{}, error){
+func (api *PublicMinerAPI) StatsGbtTxEmptyErr() {
```
Comment (dindinw) [inline services/miner/api.go:78]: should not let the method `Public` if it never has been called by the outside

File: services/miner/api.go [hunk 2]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+	api.stats.TxEmptyErrs++
```
Comment (dindinw) [inline services/miner/api.go:79]: the naming is not properly used, `NonTx` or `EmptyGbt` and It's not an error. it a status of the situation,  for example, using names like `statsEmptyGbt()` or` statsGbtNonTx()` maybe better.

File: services/miner/api.go [hunk 3]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
 	}
 }
 
+type MiningStats struct {
+	LastGBTTime          time.Time `json:"last_gbt_time"`
```
Comment (dindinw) [inline services/miner/api.go:42]: `GBT` -> `Gbt`

File: services/miner/api.go [hunk 4]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+	LastSubmit           time.Time `json:"last_submit_time"`
```
Comment (dindinw) [inline services/miner/api.go:43]: `LatestGbtSubmit`

File: services/miner/api.go [hunk 5]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+	GbtAvgTime           float64   `json:"gbt_avg_time"`
+	MaxGbtTime           float64   `json:"max_gbt_time"`
+	MaxGbtTimeLongpollid string    `json:"max_gbt_time_longpollid"`
+	MaxSubmitTime        float64   `json:"max_submit_time"`
+	MaxSubmitTimeHash    string    `json:"max_submit_time_hash"`
+	TotalGbts            int64     `json:"total_gbts"`
+	TotalGbtRequests     int64     `json:"total_gbt_requests"`
+	TotalEmptyGbts       int64     `json:"total_empty_gbts"`
+	TotalSubmits         int64     `json:"total_submits"`
+	LastTxEmptyTime      int64     `json:"-"`
```
Comment (dindinw) [inline services/miner/api.go:58]: `LastTxEmptyTime` -> `LatestMempoolEmptyDuration`

File: services/miner/api.go [hunk 6]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+	TxEmptyAvgTime       float64   `json:"tx_empty_avg_time"`
```
Comment (dindinw) [inline services/miner/api.go:59]: `MempoolEmptyAvgDuration`

File: services/miner/api.go [hunk 7]
```diff
@@ -37,32 +38,149 @@ func (m *Miner) APIs() []api.API {
...
+	TxEmptyErrs          float64   `json:"tx_empty_errs"`
```
Comment (dindinw) [inline services/miner/api.go:60]: `MempoolEmptyWarns`

File: services/miner/api.go (hunk 4 above)
Comment (dindinw) [reply services/miner/api.go:43]: LastXXX -> LatestXXX

**Review by dindinw — CHANGES_REQUESTED** [review]

File: services/miner/miner.go [hunk 8]
```diff
@@ -39,21 +39,50 @@ const (
...
+	MaxGbtRequestDuration           float64   `json:"max_gbt_request_duration"`
+	MaxGbtRequestTimeLongpollid     string    `json:"max_gbt_time_longpollid"`
+	MaxSubmitDuration               float64   `json:"max_submit_duration"`
+	MaxSubmitDurationBlockHash      string    `json:"max_submit_duration_block_hash"`
+	TotalGbts                       int64     `json:"total_gbts"`
+	TotalGbtRequests                int64     `json:"total_gbt_requests"`
+	TotalEmptyGbts                  int64     `json:"total_empty_gbts"`
+	TotalEmptyGbtRequests           int64     `json:"total_empty_request_gbts"`
+	TotalSubmits                    int64     `json:"total_submits"`
+	LastestMempoolTxEmptyDuration   int64     `json:"-"`
```
Comment (dindinw) [inline services/miner/miner.go]: `LastestMempoolTxEmptyDuration` -> `LastestMempoolEmptyDuration`
add new field `MaxMempoolEmptyDuration`

File: services/miner/miner.go [hunk 9]
```diff
@@ -74,6 +103,100 @@ type Miner struct {
 
 	RpcSer *rpc.RpcServer
 	p2pSer model.P2PService
+	stats  MiningStats
+}
+
+func (api *Miner) StatsEmptyGbt() {
```
Comment (dindinw) [inline services/miner/miner.go:109]: fix all naming : `api` -> `miner`

**Review by dindinw — CHANGES_REQUESTED** [review]

**Review by dindinw — APPROVED** [review]

**Pull request merged**